File: Spaces/ViewModels/PhotosViewModel.swift
import Combine
import FirebaseFirestore
import Foundation

@MainActor
final class PhotosViewModel: ObservableObject {
    @Published private(set) var mediaItems: [SpaceMedia]
    @Published private(set) var isLoading = false
    @Published private(set) var canUploadMedia = false
    @Published var errorMessage: String?

    let space: Space
    private let spaceService: SpaceService
    private var listener: ListenerRegistration?

    init(space: Space, mediaItems: [SpaceMedia]? = nil, spaceService: SpaceService? = nil) {
        self.space = space
        self.spaceService = spaceService ?? SpaceService()
        self.mediaItems = mediaItems ?? MockData.photosMedia(for: space)
    }

    deinit {
        listener?.remove()
    }

    func startListeningIfNeeded() {
        guard listener == nil else { return }
        isLoading = true
        Task {
            canUploadMedia = await spaceService.canPerform(.uploadPhotosVideos, in: space)
        }
        listener = spaceService.listenToMessages(in: space) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let messages):
                self.mediaItems = messages.flatMap { message in
                    message.resolvedMediaItems.compactMap { media in
                        switch media.mediaType {
                        case .photo:
                            return media.mediaCategory == "photo" ? media : nil
                        case .video:
                            return media
                        default:
                            return nil
                        }
                    }
                }
                self.isLoading = false
                self.errorMessage = nil
            case .failure(let error):
                self.mediaItems = []
                self.isLoading = false
                self.errorMessage = error.localizedDescription
            }
        }
    }

    func uploadMedia(data: Data, mimeType: String, isVideo: Bool) async {
        guard canUploadMedia else { return }
        do {
            if isVideo {
                _ = try await spaceService.sendVideoMessage(
                    in: space,
                    videoData: data,
                    caption: nil,
                    mimeType: mimeType
                )
            } else {
                _ = try await spaceService.sendImageMessage(
                    in: space,
                    imageData: data,
                    caption: nil,
                    mediaCategory: "photo"
                )
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/ui/screens/photos/PhotosViewModel.kt
package com.arcinteractive.spaces.ui.screens.photos

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcinteractive.spaces.data.model.MessageType
import com.arcinteractive.spaces.data.model.SpacePermission
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.model.SpaceMedia
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class PhotosUiState(
    val mediaItems: List<SpaceMedia>,
    val isLoading: Boolean = false,
    val lastErrorMessage: String? = null,
    val selectedMedia: SpaceMedia? = null,
    val canUploadMedia: Boolean = false
)

class PhotosViewModel(
    space: Space,
    private val spaceService: SpaceService = SpaceService()
) : ViewModel() {
    val space: Space = space
    private var listener: ListenerRegistration? = null

    private val _uiState = MutableStateFlow(
        PhotosUiState(mediaItems = emptyList())
    )
    val uiState: StateFlow<PhotosUiState> = _uiState.asStateFlow()

    fun loadPhotosIfNeeded(context: Context) {
        if (listener != null) return
        _uiState.update { it.copy(isLoading = true, lastErrorMessage = null) }
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    canUploadMedia = spaceService.canPerform(context, space, SpacePermission.UploadPhotosVideos)
                )
            }
        }
        listener = spaceService.listenToMessages(
            context = context,
            space = space,
            listenerKey = "photos.${space.id}.messages"
        ) { result ->
            result.onSuccess { messages ->
                _uiState.update {
                    it.copy(
                        mediaItems = messages.flatMap { message ->
                            message.resolvedMediaItems.mapNotNull { media ->
                                when (media.type) {
                                    MessageType.Image -> if (media.mediaCategory == "photo") media else null
                                    MessageType.Video -> media
                                    else -> null
                                }
                            }
                        },
                        isLoading = false,
                        lastErrorMessage = null
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        mediaItems = emptyList(),
                        isLoading = false,
                        lastErrorMessage = error.localizedMessage ?: "Unable to load shared photos."
                    )
                }
            }
        }
    }

    fun openMedia(media: SpaceMedia) {
        _uiState.update { it.copy(selectedMedia = media) }
    }

    fun dismissMedia() {
        _uiState.update { it.copy(selectedMedia = null) }
    }

    fun clearLastErrorMessage() {
        _uiState.update { it.copy(lastErrorMessage = null) }
    }

    fun uploadMedia(context: Context, mediaBytes: ByteArray, mimeType: String, isVideo: Boolean) {
        if (!_uiState.value.canUploadMedia) return
        viewModelScope.launch {
            runCatching {
                if (isVideo) {
                    spaceService.sendVideoMessage(
                        context = context,
                        space = space,
                        videoBytes = mediaBytes,
                        caption = null,
                        mimeType = mimeType
                    )
                } else {
                    spaceService.sendImageMessage(
                        context = context,
                        space = space,
                        imageBytes = mediaBytes,
                        caption = null,
                        mediaCategory = "photo"
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to upload this media.")
                }
            }
        }
    }

    override fun onCleared() {
        listener?.remove()
        listener = null
        super.onCleared()
    }
}
